cli/src/complete.rs

```diff
@@ -448,14 +448,27 @@ pub fn leaf_config_keys() -> Vec<CompletionCandidate> {
     config_keys_impl(true)
 }
 
-fn all_files_from_rev(rev: String) -> Vec<CompletionCandidate> {
+fn dir_prefix_from<'a>(path: &'a str, current: &str) -> Option<&'a str> {
+    path.strip_prefix(current)?
+        .split_once(std::path::MAIN_SEPARATOR)
+        .map(|(next, _)| path.split_at(current.len() + next.len() + 1).0)
+}
+
+fn all_files_from_rev(rev: String, current: &std::ffi::OsStr) -> Vec<CompletionCandidate> {
+    let Some(current) = current.to_str() else {
+        return Vec::new();
+    };
+    let cur_esc = glob::Pattern::escape(current);
     with_jj(|jj, _| {
         let mut child = jj
             .build()
             .arg("file")
             .arg("list")
             .arg("--revision")
             .arg(rev)
+            .arg("--config-toml")
+            .arg("ui.allow-filesets = true")
+            .arg(format!(r#"glob:"{cur_esc}*/**" | glob:"{cur_esc}*""#))
```

Comment:
nit: might be a bit safer to abuse string debug format to escape double quotes in `cur_esc`: `glob:{pat:?}`